parley/src/layout/data.rs

```diff
@@ -470,4 +481,43 @@ impl<B: Brush> LayoutData<B> {
             }
         }
     }
+
+    fn calculate_content_widths(&mut self) {
+        let mut max_width = 0.0;
+        for item in &self.items {
+            match item.kind {
+                LayoutItemKind::TextRun => {
+                    let run = &self.runs[item.index];
+                    let mut min_width = 0.0;
+                    let mut trailing_whitespace = 0.0;
+                    for cluster in &self.clusters[run.cluster_range.clone()] {
+                        let boundary = cluster.info.boundary();
+                        if matches!(boundary, Boundary::Line | Boundary::Mandatory) {
+                            self.min_content_width =
+                                self.min_content_width.max(min_width - trailing_whitespace);
+                            min_width = 0.0;
+                            if boundary == Boundary::Mandatory {
+                                max_width = 0.0;
+                            }
+                        }
+                        min_width += cluster.advance;
+                        trailing_whitespace = if cluster.info.whitespace().is_space_or_nbsp() {
+                            cluster.advance
+                        } else {
+                            0.0
+                        };
```

Comment:
Does this also work for RTL runs, or should those take the whitespace of the first cluster? Or perhaps the endianness should be determined based on the layout's bidi base level direction...

Reply:
Maybe add a TODO note here as a bread crumb. Parley probably needs some more tests with RTL and mixed-direction text.

Reply:
> Perhaps a premature optimisation, but could we just store the index of prev_cluster, and then compute this lazily when a line break is encoutered?

We can, and hoisting conditionals out of loops always seems sensible if possible. I'd assume this one gets compiled to a conditional move, but it's probably better not to rely on that.